components/autofill/src/encryption.rs

```diff
@@ -0,0 +1,93 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+// This is the *local* encryption support - it has nothing to do with the
+// encryption used by sync.
+
+// For context, what "local encryption" means in this context is:
+// * We use regular sqlite, but want to ensure the credit-card numbers are
+//   encrypted in the DB - so we store the number encrypted, and the key
+//   is managed by the app.
+// * The credit-card API always just accepts and returns the encrypted string,
+//   so we also expose encryption and decryption public functions that take
+//   the key and text. The core storage API never knows the unencrypted number.
+//
+// This makes life tricky for Sync - sync has its own encryption and its own
+// management of sync keys. The entire records are encrypted on the server -
+// so the record on the server has the plain-text number (which is then
+// encrypted as part of the entire record), so:
+// * When transforming a record from the DB into a Sync record, we need to
+//   *decrypt* the field.
+// * When transforming a record from Sync into a DB record, we need to *encrypt*
+//   the field.
+//
+// So Sync needs to know the key etc, and that needs to get passed down
+// multiple layers, from the app saying "sync now" all the way down to the
+// low level sync code.
+// To make life a little easier, we do that via a struct.
+
+use crate::error::*;
+
+// XXX - does this struct make sense? We could just pass the string key
+// everywhere and call public functions with 2 params?
+
+// Even though underlying primitives expect to work on bytes, keeping params
+// here as strings is convenient.
+pub(crate) struct EncryptorDecryptor {
+    key: String,
+}
+
+impl EncryptorDecryptor {
+    pub(crate) fn new(key: &str) -> Self {
+        EncryptorDecryptor {
+            key: key.to_string(),
+        }
+    }
+
+    // to be replaced with real crypto!
+    pub(crate) fn encrypt(&self, cleartext: &str) -> Result<String> {
+        Ok(self.key.clone() + cleartext) // this is secure, right? :)
+    }
+    pub(crate) fn decrypt(&self, ciphertext: &str) -> Result<String> {
+        let mut s = ciphertext.to_string();
+        let cleartext = s.split_off(self.key.len());
+        if s != self.key {
+            Err(Error::InvalidEncryptionKey)
```

Comment:
Heh, if this were real encryption, you wouldn't be able to distinguish between "wrong key" or "corrupted ciphertext", so a more generic error like `DecryptionFailure` may be more appropriate here.